### ISSUE TITLE
by_type scheduler

### ISSUE THREAD

**Issue by itsdfish** [JuliaDynamics/Agents.jl#235]
Here is my attempt to create a by_type scheduler. The main problem is that there does not appear to be an elegant way to extract types from a union. union_types was modified from a solution on stackoverflow.  

```
union_types(x::Union) = union_types(x.a, x.b)
union_types(x::Type) = (x,)
union_types(a::Type, b::Type) = (a, b)
union_types(a::Type, b::Union) = (a, union_types(b)...)
union_types(a::Union, b::Type) = (b, union_types(a)...)

function by_type(model::ABM{A}) where {A}
    types = collect(union_types(A))
    c = [Int[] for t in types]
    for a in allagents(model)
        j = findfirst(x->x == typeof(a), types)
        push!(c[j], a.id)
    end
    for i in c; shuffle!(i); end
    shuffle!(c)
    return vcat(c...)
end
```

_Originally posted by @itsdfish in https://github.com/JuliaDynamics/Agents.jl/pull/233#issuecomment-618682446_

**Comment by Libbum** [comment]
It seems pretty performant if not 'elegant'. The only other example I found in the wild is an overload of `collect`:

```julia
Base.collect(t::Union{Type, DataType, Union{}}) = _collect(t, [])
_collect(t::Type, list) = t<:Union{} ? push!(list, t) : _collect(t.b, push!(list, t.a))
_collect(t::Union{DataType,Core.TypeofBottom}, list) = push!(list, t)
```

And comparing the two:
```julia
@btime collect(Union{Wolf, Sheep, Grass})
  330.342 ns (2 allocations: 128 bytes)
 
@btime collect(union_types(Union{Wolf, Sheep, Grass}))
41.291 ns (2 allocations: 144 bytes)
```
Of course for simplicity we could overload `Base.collect` here to work directly with `union_types`.

There are some minor allocation enhancements I'm testing, but we can start a PR around this I think. We'll need to add in documentation, tests etc. An explicit mention about a convention of union order when constructing models will need to be added (if you want your sheep to move first, they must come before wolves in your union type constructor)...

Something else to think about though: this is essentially `random_by_type`. Could we alter the method in such a way that it allows us to invoke `fastest`, `by_id`, `random_activation` etc, with this method only doing the union type sorting?

**Comment by Datseris** [comment]
```julia
Base.collect(t::Union{Type, DataType, Union{}}) = _collect(t, [])
_collect(t::Type, list) = t<:Union{} ? push!(list, t) : _collect(t.b, push!(list, t.a))
_collect(t::Union{DataType,Core.TypeofBottom}, list) = push!(list, t)
```
this is **Type Piracy** and cannot be accepted in a (serious) Julia package. Type piracy is when you define a method that you do not own either of the following:

* the function the method is added to
* at least one of the types the method dispatches to

This is the case here.

**Comment by Libbum** [comment]
Good to know. It's also 7x slower. Moreso just for comparison. I haven't been able to identify why there's nothing like `union_types` defined or discussed in the documentation, nor if it's safe/captures all edge cases.

**Comment by Libbum** [comment]
Closed via #237.
